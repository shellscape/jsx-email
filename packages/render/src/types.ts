export interface RenderOptions {
  minify?: boolean;
  plainText?: boolean;
  pretty?: boolean;
  strip?: boolean;
}

export interface ProcessOptions extends Required<Omit<RenderOptions, 'plainText'>> {
  html: string;
}
