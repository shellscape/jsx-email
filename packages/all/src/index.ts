export * from '@jsx-email/body';
export * from '@jsx-email/button';
export * from '@jsx-email/column';
export * from '@jsx-email/container';
export * from '@jsx-email/font';
export * from '@jsx-email/head';
export * from '@jsx-email/heading';
export * from '@jsx-email/hr';
export * from '@jsx-email/html';
export * from '@jsx-email/img';
export * from '@jsx-email/link';
export * from '@jsx-email/markdown';
export * from '@jsx-email/preview';
export * from '@jsx-email/render';
export * from '@jsx-email/row';
export * from '@jsx-email/section';
export * from '@jsx-email/tailwind';
export * from '@jsx-email/text';
