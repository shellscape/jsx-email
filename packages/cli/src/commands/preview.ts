import { existsSync } from 'fs';
import { readFile, symlink, unlink } from 'fs/promises';
import { join, resolve } from 'path';

import chalk from 'chalk';
import globby from 'globby';
import { assert, boolean, number, object, optional, Infer } from 'superstruct';
import { createServer } from 'vite';

import { CommandFn } from './types';

const PreviewOptionsStruct = object({
  open: optional(boolean()),
  port: optional(number())
});

type PreviewOptions = Infer<typeof PreviewOptionsStruct>;

const { error } = console;

export const help = chalk`
{blue email preview}

Starts the preview server for a directory of email templates

{underline Usage}
  $ email preview <template dir path> [...options]

{underline Options}
  --no-open   Do not open a browser tab when the preview server starts
  --port      The local port number the preview server should run on. Default: 55420

{underline Examples}
  $ email preview ./src/templates --port 55420
`;

export const command: CommandFn = async (argv: PreviewOptions, input) => {
  if (input.length < 1) return false;

  assert(argv, PreviewOptionsStruct);

  const [target] = input;
  const targetPath = resolve(target);

  await start(targetPath, argv);
  return true;
};

export const start = async (targetPath: string, argv: PreviewOptions) => {
  if (!existsSync(targetPath)) {
    error(chalk`\n{red D'oh} The directory provided ({dim ${targetPath}}) doesn't exist`);
    return;
  }

  const { open = true, port = 55420 } = argv;
  const { default: config } = await import('../../app/vite.config');
  const componentPaths = await globby(join(targetPath, '/*.{jsx,tsx}'));
  const linkPath = join(config.root!, 'src/@templates');
  const templateSources = {} as Record<string, string>;

  for (const path of componentPaths) {
    templateSources[path.replace(targetPath, '').replace(/^\//, '')] =
      // eslint-disable-next-line no-await-in-loop
      await readFile(path, 'utf8');
  }

  if (existsSync(linkPath)) await unlink(linkPath);

  await symlink(targetPath, linkPath);

  const server = await createServer({
    configFile: false,
    ...config,
    define: {
      sǝɔɹnoslᴉɐɯǝxsɾ: JSON.stringify(templateSources),
      ...config.define
    },
    publicDir: targetPath,
    server: { port }
  });

  await server.listen();

  // TODO: bind CLI shortcuts. this hasn't landed in Vite yet, will be released with 5.0
  if (open) server.openBrowser();
  server.printUrls();
};
