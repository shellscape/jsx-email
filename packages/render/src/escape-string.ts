/**
 * Note: Parts of this file are derived from [React](https://github.com/facebook/react).
 * @license MIT
 */

const matchHtmlRegExp = /["'&<>]/;

const enum ASCII {
  DOUBLE_QUOTE = 34,
  AMPERSAND = 38,
  SINGLE_QUOTE = 39,
  LESS_THAN = 60,
  GREATER_THAN = 62
}

/**
 * Escapes special characters and HTML entities in a given html string.
 *
 * This is a slightly modified version of React's
 * [`escapeHtml`](https://github.com/facebook/react/blob/dddfe688206dafa5646550d351eb9a8e9c53654a/packages/react-dom-bindings/src/server/escapeTextForBrowser.js#L53)
 *
 * @param string HTML string to escape for later insertion
 */
export function escapeString(string: string) {
  const str = String(string);
  const match = matchHtmlRegExp.exec(str);

  if (!match) {
    return str;
  }

  let escape;
  let html = '';
  let index;
  let lastIndex = 0;

  for ({ index } = match; index < str.length; index++) {
    switch (str.charCodeAt(index)) {
      case ASCII.DOUBLE_QUOTE:
        escape = '&quot;';
        break;
      case ASCII.AMPERSAND:
        escape = '&amp;';
        break;
      case ASCII.SINGLE_QUOTE:
        escape = '&#x27;';
        break;
      case ASCII.LESS_THAN:
        escape = '&lt;';
        break;
      case ASCII.GREATER_THAN:
        escape = '&gt;';
        break;
      default:
        // eslint-disable-next-line no-continue
        continue;
    }

    if (lastIndex !== index) {
      html += str.slice(lastIndex, index);
    }

    lastIndex = index + 1;
    html += escape;
  }

  return lastIndex !== index ? html + str.slice(lastIndex, index) : html;
}
