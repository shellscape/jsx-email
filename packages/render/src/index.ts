export * from './render';
export * from './helpers/jsx-to-string';
