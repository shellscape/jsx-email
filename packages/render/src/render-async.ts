import { type ReadableStream } from 'node:stream/web';

import { convert } from 'html-to-text';
import pretty from 'pretty';
import { type ReactNode } from 'react';
import react from 'react-dom/server';

const { renderToStaticMarkup } = react;
const renderToStream =
  // Note: only available in platforms that support WebStreams
  // https://react.dev/reference/react-dom/server/renderToString#alternatives
  react.renderToReadableStream ||
  // Note: only available in Node
  react.renderToPipeableStream;

export const renderToString = async (children: ReactNode) => {
  const stream = await renderToStream(children);

  const html = await readableStreamToString(
    // ReactDOMServerReadableStream behaves like ReadableStream
    // in modern edge runtimes but the types are not compatible
    stream as unknown as ReadableStream<Uint8Array>
  );

  return (
    html
      // Remove leading doctype becuase we add it manually
      .replace(/^<!DOCTYPE html>/, '')
      // Remove empty comments to match the output of renderToStaticMarkup
      .replace(/<!-- -->/g, '')
  );
};

async function readableStreamToString(readableStream: ReadableStream<Uint8Array>) {
  let result = '';

  const decoder = new TextDecoder();

  for await (const chunk of readableStream) {
    result += decoder.decode(chunk);
  }

  return result;
}

export const renderAsync = async (
  component: React.ReactElement,
  options?: {
    plainText?: boolean;
    pretty?: boolean;
  }
) => {
  const markup =
    typeof renderToStaticMarkup === 'undefined'
      ? await renderToString(component)
      : renderToStaticMarkup(component);

  if (options?.plainText) {
    return convert(markup, {
      selectors: [
        { format: 'skip', selector: 'img' },
        { format: 'skip', selector: '[data-id="@jsx-email/preview"]' }
      ]
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${markup}`;

  if (options?.pretty) return pretty(document);

  return document;
};
