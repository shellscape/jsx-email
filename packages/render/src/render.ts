import { convert } from 'html-to-text';
import pretty from 'pretty';

import { jsxToString } from './jsx-to-string';

export interface Options {
  plainText?: boolean;
  pretty?: boolean;
}

export const render = (component: React.ReactElement, options?: Options) => {
  if (options?.plainText) {
    return renderAsPlainText(component, options);
  }
  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const markup = jsxToString(component);
  const document = `${doctype}${markup}`;

  if (options && options.pretty) {
    return pretty(document);
  }

  return document;
};

export const renderAsync = async (component: React.ReactElement, options?: Options) =>
  render(component, options);

const renderAsPlainText = (component: React.ReactElement, _options?: Options) =>
  convert(jsxToString(component), {
    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-id="@jsx-email/preview"]' }
    ]
  });
