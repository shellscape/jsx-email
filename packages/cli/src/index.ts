#!/usr/bin/env node
import { parseArgs } from 'node:util';

import debugConfig from 'debug';
import importLocal from 'import-local';

import pkg from '../package.json';

import { command as build } from './commands/build';
import { command as create } from './commands/create';
import { command as help } from './commands/help';
import { command as preview } from './commands/preview';
import type { CommandFn } from './commands/types';

const commands: Record<string, CommandFn> = { build, create, help, preview };
const debug = debugConfig('@jsx-email/cli');
const { log } = console;

const run = async () => {
  const argv = parseArgs({ allowPositionals: true, args: process.argv.slice(2), strict: false });
  const [commandName] = argv.positionals;
  let command = commands[commandName];

  debug(`Command Name: \`${commandName}\``);

  if (argv.values.version) {
    log(`${pkg.name} v${pkg.version}\n`);
    return;
  }

  if (!command) command = help;

  const result = await command(argv.values, argv.positionals?.slice(1) || []);

  if (!result) {
    debug(`Command \`${commandName}\` returned \`false\``);
    help({}, []);
  }
};

if (importLocal(__filename)) {
  debug('Using local install of webpack-command');
} else {
  run();
}
