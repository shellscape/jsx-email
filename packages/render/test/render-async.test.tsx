import * as React from 'react';
import ReactDOMServer from 'react-dom/server';

import { renderAsync } from '../src';

import { Preview } from './fixtures/preview';
import { Template } from './fixtures/template';

describe('renderAsync using renderToStaticMarkup', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('converts a React component into HTML', async () => {
    const actualOutput = await renderAsync(<Template firstName="Jim" />);

    expect(actualOutput).toMatchSnapshot();
  });

  it('converts a React component into PlainText', async () => {
    const actualOutput = await renderAsync(<Template firstName="Jim" />, {
      plainText: true
    });

    expect(actualOutput).toMatchSnapshot();
  });

  it('converts to plain text and removes reserved ID', async () => {
    const actualOutput = await renderAsync(<Preview />, {
      plainText: true
    });

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('renderAsync using renderToReadableStream', () => {
  const { renderToStaticMarkup } = ReactDOMServer;
  const { renderToReadableStream } = ReactDOMServer;

  const mockRenderToReadableStream = (
    component: React.ReactNode
  ): Promise<ReactDOMServer.ReactDOMServerReadableStream> => {
    const encoder = new TextEncoder();
    const markup = ReactDOMServer.renderToString(component as React.ReactElement);

    let done = false;

    return Promise.resolve({
      allReady: Promise.resolve(),
      [Symbol.asyncIterator]: () => {
        return {
          next: () => {
            if (done) {
              return Promise.resolve({ done: true, value: undefined });
            }

            done = true;

            return Promise.resolve({
              done: false,
              value: encoder.encode(markup)
            });
          }
        };
      }
    });
  };

  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();

    (ReactDOMServer.renderToStaticMarkup as unknown as undefined) = undefined;
    ReactDOMServer.renderToReadableStream = mockRenderToReadableStream;
  });

  afterEach(() => {
    ReactDOMServer.renderToStaticMarkup = renderToStaticMarkup;
    ReactDOMServer.renderToReadableStream = renderToReadableStream;
  });

  it('converts a React component into HTML', async () => {
    const actualOutput = await renderAsync(<Template firstName="Jim" />);

    expect(actualOutput).toMatchSnapshot();
  });

  it('converts a React component into PlainText', async () => {
    const actualOutput = await renderAsync(<Template firstName="Jim" />, {
      plainText: true
    });

    expect(actualOutput).toMatchSnapshot();
  });

  it('converts to plain text and removes reserved ID', async () => {
    const actualOutput = await renderAsync(<Preview />, {
      plainText: true
    });

    expect(actualOutput).toMatchSnapshot();
  });
});
