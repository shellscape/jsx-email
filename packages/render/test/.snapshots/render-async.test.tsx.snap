// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderAsync using renderToReadableStream > converts a React component into HTML 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><h1>Welcome, Jim!</h1><img src=\\"img/test.png\\" alt=\\"test\\"/><p>Thanks for trying our product. We&#x27;re thrilled to have you on board!</p>"`;

exports[`renderAsync using renderToReadableStream > converts a React component into PlainText 1`] = `
"WELCOME, JIM!

Thanks for trying our product. We're thrilled to have you on board!"
`;

exports[`renderAsync using renderToReadableStream > converts to plain text and removes reserved ID 1`] = `"THIS SHOULD BE RENDERED IN PLAIN TEXT"`;

exports[`renderAsync using renderToStaticMarkup > converts a React component into HTML 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><h1>Welcome, Jim!</h1><img src=\\"img/test.png\\" alt=\\"test\\"/><p>Thanks for trying our product. We&#x27;re thrilled to have you on board!</p>"`;

exports[`renderAsync using renderToStaticMarkup > converts a React component into PlainText 1`] = `
"WELCOME, JIM!

Thanks for trying our product. We're thrilled to have you on board!"
`;

exports[`renderAsync using renderToStaticMarkup > converts to plain text and removes reserved ID 1`] = `"THIS SHOULD BE RENDERED IN PLAIN TEXT"`;
